polyhost/services/font_downloader.py

```
@@ -0,0 +1,124 @@
+    download never leaves a truncated file that ``is_downloaded`` would trust.
+
+    ``cancel_event`` (a ``threading.Event``-like with ``is_set()``) is polled
+    between chunks; when set, the partial file is removed and ``DownloadCancelled``
+    is raised — this is how the GUI's Cancel button aborts a transfer."""
+    dest_dir = dest_dir or default_cache_dir()
+    os.makedirs(dest_dir, exist_ok=True)
+    final = os.path.join(dest_dir, font.filename)
+    if os.path.exists(final) and os.path.getsize(final) > 0:
+        return final
+    tmp = final + ".part"
+    req = urllib.request.Request(font.url, headers={"User-Agent": "PolyKybdHost"})
```

Ruff (0.15.18)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)